# Exploring Ebay Car Sales Data

**Dataquest Guided Project**

We'll work with a dataset of used cars from eBay Kleinanzeigen, a classifieds section of the German eBay website.  
The dataset was originally scraped and uploaded to Kaggle(https://www.kaggle.com/orgesleka/used-cars-database/data).  
(the version uploaded to Kaggle was cleaned to be easier to work with)

## Introduction

The aim of this project is to clean the data and analyze the included used car listings.   
The data dictionary provided with data is as follows:
- **dateCrawled** - When this ad was first crawled. All field-values are taken from this date.
- **name** - Name of the car.
- **seller** - Whether the seller is private or a dealer.
- **offerType** - The type of listing
- **price** - The price on the ad to sell the car.
- **abtest** - Whether the listing is included in an A/B test.
- **vehicleType** - The vehicle Type.
- **yearOfRegistration** - The year in which the car was first registered.
- **gearbox** - The transmission type.
- **powerPS** - The power of the car in PS.
- **model** - The car model name.
- **kilometer** - How many kilometers the car has driven.
- **monthOfRegistration** - The month in which the car was first registered.
- **fuelType** - What type of fuel the car uses.
- **brand** - The brand of the car.
- **notRepairedDamage** - If the car has a damage which is not yet repaired.
- **dateCreated** - The date on which the eBay listing was created.
- **nrOfPictures** - The number of pictures in the ad.
- **postalCode** - The postal code for the location of the vehicle.
- **lastSeenOnline** - When the crawler saw this ad last online.

In [1]:
import numpy as np
import pandas as pd

In [2]:
autos = pd.read_csv('autos.csv', encoding='latin-1')

In [3]:
autos.head()

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

In [4]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
dateCrawled            50000 non-null object
name                   50000 non-null object
seller                 50000 non-null object
offerType              50000 non-null object
price                  50000 non-null object
abtest                 50000 non-null object
vehicleType            44905 non-null object
yearOfRegistration     50000 non-null int64
gearbox                47320 non-null object
powerPS                50000 non-null int64
model                  47242 non-null object
odometer               50000 non-null object
monthOfRegistration    50000 non-null int64
fuelType               45518 non-null object
brand                  50000 non-null object
notRepairedDamage      40171 non-null object
dateCreated            50000 non-null object
nrOfPictures           50000 non-null int64
postalCode             50000 non-null int64
lastSeen               50000 non-null obj

## Cleaning Column Names

From the work we did, we can make the following observations:
- The dataset contains 20 columns, most of which are strings.
- Some columns have null values, but none have more than ~20% null values.
- The column names use camelcase instead of Python's preferred snakecase, which means we can't just replace spaces with underscores.

In [5]:
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

In [6]:
autos.columns = [
       'date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest',
       'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'nr_of_pictures', 'postal_code',
       'last_seen'
]

In [7]:
autos.head(1)

date_crawled                              name  seller offer_type  \
0  2016-03-26 17:47:46  Peugeot_807_160_NAVTECH_ON_BOARD  privat    Angebot   

    price   abtest vehicle_type  registration_year  gearbox  power_ps   model  \
0  $5,000  control          bus               2004  manuell       158  andere   

    odometer  registration_month fuel_type    brand unrepaired_damage  \
0  150,000km                   3       lpg  peugeot              nein   

            ad_created  nr_of_pictures  postal_code            last_seen  
0  2016-03-26 00:00:00               0        79588  2016-04-06 06:45:54

## Initial Exploration and Cleaning

Let's do some basic data exploration to determine what other cleaning tasks need to be done.   
Initially we will look for: 
- Text columns where all or almost all values are the same. These can often be dropped as they don't have useful information for analysis. 

- Examples of numeric data stored as text which can be cleaned and converted.

In [8]:
autos.describe()

registration_year      power_ps  registration_month  nr_of_pictures  \
count       50000.000000  50000.000000        50000.000000         50000.0   
mean         2005.073280    116.355920            5.723360             0.0   
std           105.712813    209.216627            3.711984             0.0   
min          1000.000000      0.000000            0.000000             0.0   
25%          1999.000000     70.000000            3.000000             0.0   
50%          2003.000000    105.000000            6.000000             0.0   
75%          2008.000000    150.000000            9.000000             0.0   
max          9999.000000  17700.000000           12.000000             0.0   

        postal_code  
count  50000.000000  
mean   50813.627300  
std    25779.747957  
min     1067.000000  
25%    30451.000000  
50%    49577.000000  
75%    71540.000000  
max    99998.000000

In [9]:
autos['price'].unique()

array(['$5,000', '$8,500', '$8,990', ..., '$385', '$22,200', '$16,995'],
      dtype=object)

In [10]:
autos['odometer'].unique()

array(['150,000km', '70,000km', '50,000km', '80,000km', '10,000km',
       '30,000km', '125,000km', '90,000km', '20,000km', '60,000km',
       '5,000km', '100,000km', '40,000km'], dtype=object)

In [11]:
autos['price'] = autos['price'].str.replace('$','')
autos['price'] = autos['price'].str.replace(',','')
autos['price'] = autos['price'].astype(float)
autos['odometer'] = autos['odometer'].str.replace('km','')
autos['odometer'] = autos['odometer'].str.replace(',','')
autos['odometer'] = autos['odometer'].astype(float)
autos.rename(columns={'odometer':'odometer_km'}, inplace=True)

In [12]:
autos.head(1)

date_crawled                              name  seller offer_type  \
0  2016-03-26 17:47:46  Peugeot_807_160_NAVTECH_ON_BOARD  privat    Angebot   

    price   abtest vehicle_type  registration_year  gearbox  power_ps   model  \
0  5000.0  control          bus               2004  manuell       158  andere   

   odometer_km  registration_month fuel_type    brand unrepaired_damage  \
0     150000.0                   3       lpg  peugeot              nein   

            ad_created  nr_of_pictures  postal_code            last_seen  
0  2016-03-26 00:00:00               0        79588  2016-04-06 06:45:54

## Exploring the Odometer and Price Columns

We'll start by analyzing the odometer_km and price columns.Here's the steps we'll take:  
- Analyze the columns using minimum and maximum values and look for any values that look unrealistically high or low (outliers) that we might want to remove.
- We'll use:
 - **Series.unique().shape** to see how many unique values
 - **Series.describe()** to view min/max/median/mean etc
 - **Series.value_counts()**, with some variations:
   - chained to **.head()** if there are lots of values.
   - Because **Series.value_counts()** returns a series, we can use **Series.sort_index()** with **ascending=** **True** or **False** to view the highest and lowest values with their counts (can also chain to **head()** here).
 - When removing outliers, we can do **df[(df["col"] > x ) & (df["col"] < y )]**, but it's more readable to use **df[df["col"].between(x,y)]**

In [13]:
autos['odometer_km'].unique().shape

(13,)

In [14]:
autos['odometer_km'].describe() 

count     50000.000000
mean     125732.700000
std       40042.211706
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64

In [15]:
autos['odometer_km'].value_counts()

150000.0    32424
125000.0     5170
100000.0     2169
90000.0      1757
80000.0      1436
70000.0      1230
60000.0      1164
50000.0      1027
5000.0        967
40000.0       819
30000.0       789
20000.0       784
10000.0       264
Name: odometer_km, dtype: int64

In [16]:
autos['price'].unique().shape

(2357,)

In [17]:
autos['price'].describe() 

count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64

In [18]:
autos['price'].value_counts().head()

0.0       1421
500.0      781
1500.0     734
2500.0     643
1200.0     639
Name: price, dtype: int64

## Exploring the date columns

Let's now move on to the date columns and understand the date range the data covers.

There are 5 columns that should represent date values. Some of these columns were created by the crawler, some came from the website itself. We can differentiate by referring to the data dictionary:
- **date_crawled**: added by the crawler
- **last_seen**: added by the crawler
- **ad_created**: from the website
- **registration_month**: from the website
- **registration_year**: from the website

Right now, the **date_crawled**, **last_seen**, and **ad_created** columns are all identified as string values by pandas. Because these three columns are represented as strings, we need to convert the data into a numerical representation so we can understand it quantitatively. The other two columns are represented as numeric values, so we can use methods like **Series.describe()** to understand the distribution without any extra data processing.

Let's first understand how the values in the three string columns are formatted. These columns all represent full timestamp values, like so:

In [19]:
autos[['date_crawled','ad_created','last_seen']][0:5]

date_crawled           ad_created            last_seen
0  2016-03-26 17:47:46  2016-03-26 00:00:00  2016-04-06 06:45:54
1  2016-04-04 13:38:56  2016-04-04 00:00:00  2016-04-06 14:45:08
2  2016-03-26 18:57:24  2016-03-26 00:00:00  2016-04-06 20:15:37
3  2016-03-12 16:58:10  2016-03-12 00:00:00  2016-03-15 03:16:28
4  2016-04-01 14:38:50  2016-04-01 00:00:00  2016-04-01 14:38:50

You'll notice that the first 10 characters represent the day (e.g. **2016-03-12**). To understand the date range, we can extract just the date values, use **Series.value_counts()** to generate a distribution, and then sort by the index.

To select the first 10 characters in each column, we can use **Series.str[:10]**:

In [20]:
autos['date_crawled'].str[:10]

0        2016-03-26
1        2016-04-04
2        2016-03-26
3        2016-03-12
4        2016-04-01
5        2016-03-21
6        2016-03-20
7        2016-03-16
8        2016-03-22
9        2016-03-16
10       2016-03-15
11       2016-03-16
12       2016-03-31
13       2016-03-23
14       2016-03-23
15       2016-04-01
16       2016-03-16
17       2016-03-29
18       2016-03-26
19       2016-03-17
20       2016-03-05
21       2016-03-06
22       2016-03-28
23       2016-03-10
24       2016-04-03
25       2016-03-21
26       2016-04-03
27       2016-03-27
28       2016-03-19
29       2016-04-02
            ...    
49970    2016-03-21
49971    2016-03-29
49972    2016-03-26
49973    2016-03-27
49974    2016-03-20
49975    2016-03-27
49976    2016-03-19
49977    2016-03-31
49978    2016-04-04
49979    2016-03-20
49980    2016-03-12
49981    2016-03-15
49982    2016-03-29
49983    2016-03-06
49984    2016-03-31
49985    2016-04-02
49986    2016-04-04
49987    2016-03-22
49988    2016-03-28


In [21]:
autos['date_crawled'].value_counts(normalize=True, dropna=False)

2016-03-27 22:55:05    0.00006
2016-03-21 16:37:21    0.00006
2016-03-14 20:50:02    0.00006
2016-03-25 19:57:10    0.00006
2016-03-23 18:39:34    0.00006
2016-03-29 23:42:13    0.00006
2016-03-23 19:38:20    0.00006
2016-03-21 20:37:19    0.00006
2016-03-11 22:38:16    0.00006
2016-03-22 09:51:06    0.00006
2016-03-10 15:36:24    0.00006
2016-03-05 16:57:05    0.00006
2016-04-02 11:37:04    0.00006
2016-03-08 10:40:35    0.00006
2016-04-02 15:49:30    0.00006
2016-03-19 17:36:18    0.00006
2016-04-04 16:40:33    0.00006
2016-03-09 11:54:38    0.00006
2016-03-12 16:06:22    0.00006
2016-03-30 17:37:35    0.00006
2016-03-30 19:48:02    0.00006
2016-03-16 21:50:53    0.00006
2016-03-09 23:25:18    0.00004
2016-03-23 20:25:18    0.00004
2016-04-04 14:48:38    0.00004
2016-03-30 02:57:54    0.00004
2016-04-02 14:58:27    0.00004
2016-03-11 10:54:11    0.00004
2016-03-20 15:45:40    0.00004
2016-03-20 12:25:21    0.00004
                        ...   
2016-03-25 20:49:27    0.00002
2016-03-

In [22]:
autos['date_crawled'].sort_index()

0        2016-03-26 17:47:46
1        2016-04-04 13:38:56
2        2016-03-26 18:57:24
3        2016-03-12 16:58:10
4        2016-04-01 14:38:50
5        2016-03-21 13:47:45
6        2016-03-20 17:55:21
7        2016-03-16 18:55:19
8        2016-03-22 16:51:34
9        2016-03-16 13:47:02
10       2016-03-15 01:41:36
11       2016-03-16 18:45:34
12       2016-03-31 19:48:22
13       2016-03-23 10:48:32
14       2016-03-23 11:50:46
15       2016-04-01 12:06:20
16       2016-03-16 14:59:02
17       2016-03-29 11:46:22
18       2016-03-26 19:57:44
19       2016-03-17 13:36:21
20       2016-03-05 19:57:31
21       2016-03-06 19:07:10
22       2016-03-28 20:50:54
23       2016-03-10 19:55:34
24       2016-04-03 11:57:02
25       2016-03-21 21:56:18
26       2016-04-03 22:46:28
27       2016-03-27 18:45:01
28       2016-03-19 21:56:19
29       2016-04-02 12:45:44
                ...         
49970    2016-03-21 22:47:37
49971    2016-03-29 14:54:12
49972    2016-03-26 22:25:23
49973    2016-

In [23]:
autos['ad_created'].value_counts(normalize=True, dropna=False)

2016-04-03 00:00:00    0.03892
2016-03-20 00:00:00    0.03786
2016-03-21 00:00:00    0.03772
2016-04-04 00:00:00    0.03688
2016-03-12 00:00:00    0.03662
2016-03-14 00:00:00    0.03522
2016-04-02 00:00:00    0.03508
2016-03-28 00:00:00    0.03496
2016-03-07 00:00:00    0.03474
2016-03-29 00:00:00    0.03414
2016-03-19 00:00:00    0.03384
2016-04-01 00:00:00    0.03380
2016-03-15 00:00:00    0.03374
2016-03-30 00:00:00    0.03344
2016-03-08 00:00:00    0.03334
2016-03-09 00:00:00    0.03324
2016-03-22 00:00:00    0.03280
2016-03-11 00:00:00    0.03278
2016-03-26 00:00:00    0.03256
2016-03-23 00:00:00    0.03218
2016-03-31 00:00:00    0.03192
2016-03-25 00:00:00    0.03188
2016-03-10 00:00:00    0.03186
2016-03-17 00:00:00    0.03120
2016-03-27 00:00:00    0.03090
2016-03-16 00:00:00    0.03000
2016-03-24 00:00:00    0.02908
2016-03-05 00:00:00    0.02304
2016-03-13 00:00:00    0.01692
2016-03-06 00:00:00    0.01512
                        ...   
2016-02-12 00:00:00    0.00006
2016-02-

In [24]:
autos['ad_created'].sort_index()

0        2016-03-26 00:00:00
1        2016-04-04 00:00:00
2        2016-03-26 00:00:00
3        2016-03-12 00:00:00
4        2016-04-01 00:00:00
5        2016-03-21 00:00:00
6        2016-03-20 00:00:00
7        2016-03-16 00:00:00
8        2016-03-22 00:00:00
9        2016-03-16 00:00:00
10       2016-03-14 00:00:00
11       2016-03-16 00:00:00
12       2016-03-31 00:00:00
13       2016-03-23 00:00:00
14       2016-03-23 00:00:00
15       2016-04-01 00:00:00
16       2016-03-16 00:00:00
17       2016-03-29 00:00:00
18       2016-03-26 00:00:00
19       2016-03-17 00:00:00
20       2016-03-05 00:00:00
21       2016-03-06 00:00:00
22       2016-03-28 00:00:00
23       2016-03-10 00:00:00
24       2016-04-03 00:00:00
25       2016-03-21 00:00:00
26       2016-04-03 00:00:00
27       2016-03-27 00:00:00
28       2016-03-19 00:00:00
29       2016-04-02 00:00:00
                ...         
49970    2016-03-21 00:00:00
49971    2016-03-29 00:00:00
49972    2016-03-26 00:00:00
49973    2016-

In [25]:
autos['last_seen'].value_counts(normalize=True, dropna=False)

2016-04-07 06:17:27    0.00016
2016-04-06 06:17:24    0.00014
2016-04-06 21:17:51    0.00014
2016-04-07 03:16:17    0.00014
2016-04-06 01:16:01    0.00012
2016-04-06 02:17:26    0.00012
2016-04-06 05:16:14    0.00012
2016-04-06 14:17:04    0.00012
2016-04-06 22:17:26    0.00012
2016-04-05 16:44:47    0.00012
2016-04-06 07:46:11    0.00012
2016-04-06 07:46:03    0.00012
2016-04-07 05:16:17    0.00012
2016-04-07 04:46:51    0.00012
2016-04-06 19:16:38    0.00012
2016-04-06 15:17:56    0.00012
2016-04-06 02:16:12    0.00012
2016-04-06 10:17:12    0.00012
2016-04-07 03:45:23    0.00012
2016-04-06 19:15:21    0.00012
2016-04-06 15:45:50    0.00012
2016-04-06 08:44:19    0.00012
2016-04-06 14:44:55    0.00012
2016-04-06 15:16:45    0.00012
2016-04-06 13:17:03    0.00012
2016-04-07 06:46:12    0.00012
2016-04-06 20:48:27    0.00012
2016-04-07 04:46:40    0.00010
2016-04-07 07:44:37    0.00010
2016-04-06 08:44:40    0.00010
                        ...   
2016-03-12 02:16:44    0.00002
2016-04-

In [26]:
autos['last_seen'].sort_index()

0        2016-04-06 06:45:54
1        2016-04-06 14:45:08
2        2016-04-06 20:15:37
3        2016-03-15 03:16:28
4        2016-04-01 14:38:50
5        2016-04-06 09:45:21
6        2016-03-23 02:48:59
7        2016-04-07 03:17:32
8        2016-03-26 18:18:10
9        2016-04-06 10:46:35
10       2016-04-07 03:16:21
11       2016-03-16 18:45:34
12       2016-04-06 14:17:52
13       2016-04-01 14:17:13
14       2016-04-01 15:46:47
15       2016-04-02 21:10:48
16       2016-03-18 05:29:37
17       2016-04-05 20:46:26
18       2016-04-02 12:16:41
19       2016-03-17 14:45:58
20       2016-03-07 05:46:46
21       2016-04-05 23:46:19
22       2016-04-01 06:45:30
23       2016-03-17 08:45:17
24       2016-04-07 13:16:50
25       2016-04-01 05:16:49
26       2016-04-05 23:46:48
27       2016-03-27 18:45:01
28       2016-04-07 14:58:48
29       2016-04-02 12:45:44
                ...         
49970    2016-04-07 04:17:34
49971    2016-03-29 20:41:51
49972    2016-03-28 11:28:18
49973    2016-

In [27]:
autos['registration_year'].describe()

count    50000.000000
mean      2005.073280
std        105.712813
min       1000.000000
25%       1999.000000
50%       2003.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

## Dealing with Incorrect Registration Year Data

One thing that stands out from the exploration we did in the last screen is that the registration_year column contains some odd values:
- The minimum value is **1000**, before cars were invented
- The maximum value is **9999**, many years into the future

Because a car can't be first registered after the listing was seen, any vehicle with a registration year above 2016 is definitely inaccurate. Determining the earliest valid year is more difficult. Realistically, it could be somewhere in the first few decades of the 1900s.

Let's count the number of listings with cars that fall outside the 1900 - 2016 interval and see if it's safe to remove those rows entirely, or if we need more custom logic.

In [28]:
autos.loc[~ autos['registration_year'].between(1899, 2017), 'registration_year'].shape

(519,)

In [29]:
autos.loc[~ autos['registration_year'].between(1899, 2017), 'registration_year'] = np.nan

In [30]:
autos.dropna(subset=['registration_year'], axis=0, inplace=True)

In [31]:
autos['registration_year'].value_counts(normalize=True)

2000.0    0.067784
2005.0    0.060932
1999.0    0.060629
2004.0    0.055314
2003.0    0.055112
2006.0    0.054728
2001.0    0.054627
2002.0    0.051191
1998.0    0.049575
2007.0    0.046563
2008.0    0.045088
2009.0    0.042400
1997.0    0.040985
2011.0    0.033023
2010.0    0.032275
2017.0    0.029365
1996.0    0.029183
2012.0    0.026738
2016.0    0.026596
1995.0    0.026535
2013.0    0.016289
2014.0    0.013460
1994.0    0.013338
1993.0    0.008993
2015.0    0.008064
1990.0    0.007983
1992.0    0.007902
1991.0    0.007195
1989.0    0.003658
1988.0    0.002870
            ...   
1977.0    0.000445
1966.0    0.000445
1975.0    0.000384
1969.0    0.000384
1965.0    0.000344
1964.0    0.000243
1910.0    0.000182
1963.0    0.000182
1959.0    0.000141
1961.0    0.000121
1956.0    0.000101
1962.0    0.000081
1958.0    0.000081
1937.0    0.000081
1950.0    0.000061
1934.0    0.000040
1941.0    0.000040
1951.0    0.000040
1954.0    0.000040
1955.0    0.000040
1957.0    0.000040
1952.0    0.

In [32]:
autos.shape

(49481, 20)

## Exploring Price by Brand

When working with data on cars, it's natural to explore variations across different car brands. We can use aggregation to understand the **brand** column. Here's what the process looks like:  
- Identify the unique values we want to aggregate by
- Create an empty dictionary to store our aggregate data
- Loop over the unique values, and for each:
    - Subset the dataframe by the unique values
    - Calculate the mean of whichever column we're interested in
    - Assign the val/mean to the dict as k/v.

In [33]:
autos['brand'].unique()

array(['peugeot', 'bmw', 'volkswagen', 'smart', 'ford', 'chrysler',
       'seat', 'renault', 'mercedes_benz', 'audi', 'sonstige_autos',
       'opel', 'mazda', 'porsche', 'mini', 'toyota', 'dacia', 'nissan',
       'jeep', 'saab', 'volvo', 'mitsubishi', 'jaguar', 'fiat', 'skoda',
       'subaru', 'kia', 'citroen', 'chevrolet', 'hyundai', 'honda',
       'daewoo', 'suzuki', 'trabant', 'land_rover', 'alfa_romeo', 'lada',
       'rover', 'daihatsu', 'lancia'], dtype=object)

In [34]:
autos['brand'].value_counts()

volkswagen        10566
opel               5391
bmw                5383
mercedes_benz      4690
audi               4254
ford               3451
renault            2368
peugeot            1449
fiat               1288
seat                915
skoda               782
mazda               751
nissan              749
smart               695
citroen             693
toyota              615
sonstige_autos      534
hyundai             485
volvo               454
mini                424
mitsubishi          399
honda               391
kia                 353
alfa_romeo          327
porsche             293
suzuki              289
chevrolet           279
chrysler            181
dacia               128
daihatsu            126
jeep                109
subaru              107
land_rover           99
saab                 78
daewoo               78
trabant              77
jaguar               76
rover                69
lancia               56
lada                 29
Name: brand, dtype: int64

In [35]:
brand_mean_prices = {}
brands = autos['brand'].unique()

for brand in brands:
    brand_mean_prices[brand] = autos.loc[autos['brand'] == brand, 'price'].mean()

In [36]:
brand_mean_prices

{'alfa_romeo': 3955.4495412844035,
 'audi': 8989.552891396334,
 'bmw': 8279.482258963404,
 'chevrolet': 6496.311827956989,
 'chrysler': 3286.0552486187844,
 'citroen': 43097.88888888889,
 'dacia': 5920.3828125,
 'daewoo': 1049.1025641025642,
 'daihatsu': 1535.4603174603174,
 'fiat': 12276.558229813665,
 'ford': 7140.0625905534625,
 'honda': 3928.2762148337597,
 'hyundai': 5303.065979381443,
 'jaguar': 11176.197368421053,
 'jeep': 11377.550458715596,
 'kia': 5724.385269121813,
 'lada': 2502.896551724138,
 'lancia': 3057.339285714286,
 'land_rover': 18934.272727272728,
 'mazda': 3968.7709720372836,
 'mercedes_benz': 29704.757995735607,
 'mini': 10392.393867924528,
 'mitsubishi': 3320.187969924812,
 'nissan': 4589.3564753004,
 'opel': 5143.339454646633,
 'peugeot': 3014.3381642512077,
 'porsche': 44553.46757679181,
 'renault': 2364.367820945946,
 'rover': 1494.5217391304348,
 'saab': 3192.269230769231,
 'seat': 4259.546448087432,
 'skoda': 6300.735294117647,
 'smart': 3494.264748201439,
 

## Storing Aggregate Data in a DataFrame

we aggregated across brands to understand mean price. We observed that in the top 6 brands, there's a distinct price gap.
- Audi, BMW and Mercedes Benz are more expensive
- Ford and Opel are less expensive
- Volkswagen is in between

For the top 6 brands, let's use aggregation to understand the average mileage for those cars and if there's any visible link with mean price. While our natural instinct may be to display both aggregated series objects and visually compare them, this has a few limitations:

- it's difficult to compare more than two aggregate series objects if we want to extend to more columns
- we can't compare more than a few rows from each series object
- we can only sort by the index (brand name) of both series objects so we can easily make visual comparisons

Instead, we can combine the data from both series objects into a single dataframe (with a shared index) and display the dataframe directly. To do this, we'll need to learn two pandas methods:
- pandas series constructor
- pandas dataframe constructor

In [37]:
brand_mean_mileage = {}
for brand in brands:
    brand_mean_mileage[brand] = autos.loc[autos['brand'] == brand, 'odometer_km'].mean()

In [38]:
brand_mean_mileage

{'alfa_romeo': 131437.30886850154,
 'audi': 129561.58909261871,
 'bmw': 132520.89912688092,
 'chevrolet': 99014.33691756273,
 'chrysler': 133149.17127071825,
 'citroen': 119834.05483405484,
 'dacia': 84218.75,
 'daewoo': 121346.15384615384,
 'daihatsu': 115714.28571428571,
 'fiat': 116836.1801242236,
 'ford': 124058.24398725007,
 'honda': 123631.71355498722,
 'hyundai': 106690.72164948453,
 'jaguar': 120921.05263157895,
 'jeep': 127522.93577981651,
 'kia': 112719.54674220963,
 'lada': 85517.24137931035,
 'lancia': 123125.0,
 'land_rover': 118333.33333333333,
 'mazda': 124933.42210386152,
 'mercedes_benz': 131041.57782515991,
 'mini': 89375.0,
 'mitsubishi': 126904.76190476191,
 'nissan': 118951.93591455274,
 'opel': 129326.65553700613,
 'peugeot': 127311.939268461,
 'porsche': 97457.33788395904,
 'renault': 128230.57432432432,
 'rover': 136449.27536231885,
 'saab': 144487.1794871795,
 'seat': 122027.32240437159,
 'skoda': 111074.16879795397,
 'smart': 100489.20863309353,
 'sonstige_aut

In [39]:
bmp = pd.Series(brand_mean_prices)
bmm = pd.Series(brand_mean_mileage)
bm = pd.DataFrame(bmp, columns=['mean_prices'])
bm['mean_mileage'] = bmm

In [40]:
bm

mean_prices   mean_mileage
alfa_romeo       3955.449541  131437.308869
audi             8989.552891  129561.589093
bmw              8279.482259  132520.899127
chevrolet        6496.311828   99014.336918
chrysler         3286.055249  133149.171271
citroen         43097.888889  119834.054834
dacia            5920.382812   84218.750000
daewoo           1049.102564  121346.153846
daihatsu         1535.460317  115714.285714
fiat            12276.558230  116836.180124
ford             7140.062591  124058.243987
honda            3928.276215  123631.713555
hyundai          5303.065979  106690.721649
jaguar          11176.197368  120921.052632
jeep            11377.550459  127522.935780
kia              5724.385269  112719.546742
lada             2502.896552   85517.241379
lancia           3057.339286  123125.000000
land_rover      18934.272727  118333.333333
mazda            3968.770972  124933.422104
mercedes_benz   29704.757996  131041.577825
mini            10392.393868   89375.000000
mitsubishi       3320.187970  126904.761905
nissan           4589.356475  118951.935915
opel             5143.339455  129326.655537
peugeot          3014.338164  127311.939268
porsche         44553.467577   97457.337884
renault          2364.367821  128230.574324
rover            1494.521739  136449.275362
saab             3192.269231  144487.179487
seat             4259.546448  122027.322404
skoda            6300.735294  111074.168798
smart            3494.264748  100489.208633
sonstige_autos  39098.705993   87303.370787
subaru           3702.140187  125327.102804
suzuki           4028.591696  109290.657439
toyota           5101.678049  116105.691057
trabant          1557.207792   60064.935065
volkswagen       6417.041170  128976.433844
volvo            4690.112335  138557.268722

## Next Steps

In this guided project, we practiced applying a variety of pandas methods to explore and understand a data set on car listings. Here are some next steps for you to consider:
- Data cleaning next steps:
  - Identify categorical data that uses german words, translate them and map the values to their english counterparts
  - Convert the dates to be uniform numeric data, so **"2016-03-21"** becomes the integer **20160321**.
  - See if there are particular keywords in the name column that you can extract as new columns
- Analysis next steps:
  - Find the most common brand/model combinations
  - Split the **odometer_km** into groups, and use aggregation to see if average prices follows any patterns based on the milage.
  - How much cheaper are cars with damage than their non-damaged counterparts?